honeybadgermpc/hbavss.py

```diff
@@ -323,28 +270,156 @@ def _get_dealer_msg(self, values):
+
+        # verify & send all ok
+        for k in range(secret_size):
+            if not self.poly_commit.verify_eval(
+                    commitments[k], self.my_id+1, shares[k], witnesses[k]):
+                # invalid share found -> sendall (IMPLICATE,SKi,k)
+                multicast((HbAVSSMessageType.IMPLICATE, self.private_key, k))
+
+        multicast((HbAVSSMessageType.OK, ""))
+
+        ok_set = set()
+        ready_set = set()
+        r1_set = set()
+        r2_set = set()
+        ready_sent = False
+        r1_sent = False
+        r2_sent = False
+        r1_phi = [None] * self.n
+        r2_phi = [None] * self.n
+
+        while True:
+            # main recv loop for Bracha-style agreement and implicate handling
+            sender, avss_msg = await recv()
+            # OK
+            if avss_msg[0] == HbAVSSMessageType.OK and sender not in ok_set:
+                ok_set.add(sender)
+            # READY
+            if avss_msg[0] == HbAVSSMessageType.READY and sender not in ready_set:
+                ready_set.add(sender)
+            # IMPLICATE
+            if avss_msg[0] == HbAVSSMessageType.IMPLICATE:
+                # Handling Implication
+                j = sender
+                j_pk = avss_msg[1]
+                j_k = avss_msg[2]
+                # discard if PKj ! = g^SKj
+                if self.public_keys[j] != pow(self.g, j_pk):
+                    logger.debug("[%d] invalid implicate from sender %s",
+                                 self.my_id, sender)
+                    continue
+                # decrypt and verify
+                implicate_msg = await avid.retrieve(tag, j)
+                j_z = loads(implicate_msg)
+                j_shared_key = pow(ephemeral_public_key, j_pk)
+                j_share, j_witnesses = SymmetricCrypto.decrypt(
+                    str(j_shared_key).encode("utf-8"), j_z[j_k])
+                if self.poly_commit.verify_eval(
+                        commitments[j_k], sender+1, j_share, j_witnesses):
+                    continue
+                # share recovery
+                logger.debug("[%d] Share recovery activated by %d", self.my_id, j)
+                # if my shares are valid
+                if not r1_sent and self.shares_future.done():
+                    r1_sent = True
+                    for j in range(self.n):
+                        phi_i_j = polynomials_over(ZR).interpolate_at(
+                            zip(range(0, secret_size), shares), x_recomb=j)
+                        w_i_j = polynomials_over(ZR).interpolate_at(
+                            zip(range(0, secret_size), witnesses), x_recomb=j)
+                        send(j, (HbAVSSMessageType.RECOVERY1, phi_i_j, w_i_j))
+            # R1
+            if avss_msg[0] == HbAVSSMessageType.RECOVERY1:
+                _, phi_k_i, w_k_i = avss_msg
+                # Will be uncommented once the interpolation of commitments is done
```

Comment:
@tyurek said he'll implement the interpolation on the commitments. So, I commented this part out for now.